light-system-programs/tests/user-random.tests.ts

```diff
@@ -0,0 +1,473 @@
+      relayerKeypair.publicKey,
+      LOOK_UP_TABLE,
+      relayerRecipientSol.publicKey,
+      new BN(100000),
+      new BN(10_100_000),
+      relayerKeypair,
+    );
+
+    let testUsers: { user: User; wallet: Keypair }[] = [];
+    for (let user = 0; user < noUsers; user++) {
+      testUsers.push(await createTestUser(anchorProvider, relayer));
+    }
+
+    /**
+     * select user
+     * select action
+     * perform action
+     */
+    let transactions = 0;
+    while (true) {
+      console.log("\n----------------------------------\n");
+      const res = getRandomObjectAndArray(testUsers);
+      const rndUser = res[0].user;
+      const wallet = res[0].wallet;
+      const remainingTestUsers = res[1];
+      const _token = token != "BOTH" ? token : getRandomObject(["SOL", "USDC"]);
+      const tokenMint = TOKEN_REGISTRY.get(_token).mint;
+      console.log("no users ", testUsers.length);
+      console.log("no transactions ", transactions);
+      await rndUser.getBalance();
+      const randomAction = await selectRandomAction(rng, rndUser, tokenMint);
+      console.log("random action ", randomAction);
+      console.log("selected user ", rndUser.account.getPublicKey());
+      console.log("selected _token ", _token);
+
+      try {
+        if (randomAction === Action.SHIELD) {
+          await mergeAllInboxUtxos(rndUser, _token);
+          await shield(rndUser, rng, true, _token, wallet);
+        } else if (randomAction === Action.TRANSFER) {
+          await mergeAllInboxUtxos(rndUser, _token);
+          const createNewUser = rng();
+          let recipientUser;
+          if (remainingTestUsers.length == 0 || createNewUser < 0.5 / noUsers) {
+            recipientUser = await createTestUser(anchorProvider, relayer);
+            testUsers.push(recipientUser);
+          } else {
+            recipientUser = getRandomObject(remainingTestUsers);
+          }
+          console.log(
+            "selected recipientUser ",
+            recipientUser.user.account.getPublicKey(),
+          );
+
+          await transfer(rndUser, recipientUser.user, rng, true, _token);
+        } else {
+          await mergeAllInboxUtxos(rndUser, _token);
+          await unshield(rndUser, rng, true, _token);
+        }
+        transactions++;
+      } catch (error) {
+        const inboxBalance = await rndUser.getUtxoInbox();
+        console.log("inboxBalance ", inboxBalance);
+        const balance = await rndUser.getBalance();
+        console.log("balance ", balance);
+        throw error;
+      }
+    }
+  }
+
+  it.skip("random sol test", async () => {
+    await randomTest(1, anchorProvider);
+  });
+  it.skip("random spl test", async () => {
+    await randomTest(1, anchorProvider, "USDC");
+  });
+
+  /**
+   * fails at tx 22
+   * - had two shields before one usdc and one sol
+   * - history only finds the lastest sol shield
+   *
+   * Check:
+   * - that usdc transactions are categorized correctly
+   * */
```

Comment:
I suppose this is wip?

Reply:
yes, and the tx history is out of scope for this pr